Experimental: added LTOR query expansion technique in rephrase_using_translation

karthikRavichandran opened the pull request:
- [x] **PR title**: "Experimental: added LTOR query expansion technique in rephrase_using_translation "
- [x] **PR message**: 
    - **Description:** Adding new query expansion technique in retriever that uses multilevel translation  
    - **Issue:** the issue # Please refer https://arxiv.org/abs/2402.13500v1  for method and initial results
    - **Dependencies:** Users have to install torch and transformers while using this functionality 
    - **Twitter handle:** [@KARTHIKRAVI007](https://twitter.com/KARTHIKRAVI007)
- [x] **Add tests and docs**: 
  1. The simple notebook is placed in `docs/docs/integrations/retrievers/rephrase_using_translation.ipynb`

karthikRavichandran committed:
[LTOR_retriver](https://arxiv.org/abs/2402.13500) is a technique used in this paper to Tailoring LLM personalization ith user profiles. However, this can also be used in normal retrivel task. This work expands the queries with the help of multilevel translation. Except translation, it uses basic BM25 algorithm
    "\n",

karthikRavichandran committed:
Merge branch 'langchain-ai:master' into master

karthikRavichandran committed:
Cleared Jupyter notebook's redundant outputs

karthikRavichandran committed:
Cleared Jupyter notebook's redundant outputs

karthikRavichandran committed:
Merge remote-tracking branch 'origin/master'

karthikRavichandran committed:
Cleared Jupyter notebook's redundant outputs

karthikRavichandran committed:
added TranslateQueryExpansionRetriever in __init__.py

karthikRavichandran committed:
minor spelling fix

karthikRavichandran commented:
Hi @baskaryan, @efriis, @eyurtsev, @hwchase17, Can any of you please look at my pull request? I don't know why only Vercel succeeded and the other three have been running since the creation of pull request.

hwchase17 commented:
this is really interesting, but implementing it as a retriever is not quite the right way

I would take the logic inside rephrase_using_translation, add it to langchain-experimental as a standalone class, and then show how to use (the notebook you added is fine)

karthikRavichandran committed:
Merge branch 'master' into master

karthikRavichandran committed:
Merge branch 'langchain-ai:master' into master

karthikRavichandran commented on libs/langchain/langchain/retrievers/ltor_retriever.py:
```diff
@@ -0,0 +1,6 @@
+from langchain_community.retrievers.ltor_retriever import (
```
Sure. I will remove these imports. Also, I'll update the main class and refactore it as you mentioned in the previous comment.

karthikRavichandran committed:
Merge branch 'langchain-ai:master' into master

karthikRavichandran committed:
Converted LTOR to rephrase_using_translation and added a directory in experimental

karthikRavichandran committed:
added rephrase_using_translation notebook.

karthikRavichandran committed:
Merge remote-tracking branch 'origin/master'

karthikRavichandran commented:
@hwchase17 I made the suggested changes: added a directory rephrase_using_translation in the experimental and added the file rephrase_using_translation.py which has LTORQueryExpansion class. I have also made changes to the Python Notebook. Please go review my changes and let me know if I have to modify the structure or name of the class further.

karthikRavichandran committed:
Fixed lint errors

karthikRavichandran committed:
Merge branch 'master' into master

karthikRavichandran committed:
Merge branch 'master' into master

karthikRavichandran committed:
Fixed lint errors and add doc string.

karthikRavichandran committed:
Merge remote-tracking branch 'origin/master'

karthikRavichandran committed:
Merge branch 'master' into master

karthikRavichandran commented:
@hwchase17 I made the suggested changes but a few pylint tests are failing. I couldn't understand this error "I001 Import block is un-sorted or un-formatted" for ___init__.py file, as this file has only one import.

baskaryan commented:
> @hwchase17 I made the suggested changes but a few pylint tests are failing. I couldn't understand this error "I001 Import block is un-sorted or un-formatted" for _**init**.py file, as this file has only one import.

to format and lint locally run
```bash
cd libs/experimental
poetry install --with lint,test,typing
make format
make lint
```

karthikRavichandran committed:
Fixed pylint issue by renaming rephrase_using_translation.py to ltor_query_expansion.py. The issue I faced with "langchain_experimental.rephrase_using_translation.rephrase_using_translation" is exceeding 88 char in a line. `make format` will not help because of https://github.com/astral-sh/ruff/issues/4894

karthikRavichandran committed:
Merge remote-tracking branch 'origin/master'

karthikRavichandran commented:
> > @hwchase17 I made the suggested changes but a few pylint tests are failing. I couldn't understand this error "I001 Import block is un-sorted or un-formatted" for _**init**.py file, as this file has only one import.
> 
> to format and lint locally run
> 
> ```shell
> cd libs/experimental
> poetry install --with lint,test,typing
> make format
> make lint
> ```
@baskaryan 
Thanks for the help. I have made the respective changes and pushed the code. Here, after running `make format`, it formatted the __init__.py, but the changes it made still exceeded the 88 char in a line. It seems there is no solution for this, please check this (https://github.com/astral-sh/ruff/issues/4894). So, I renamed the file to keep the import (at  __init__.py) in Pylint's 88  char line constraint.

karthikRavichandran committed:
Merge branch 'master' into master

karthikRavichandran committed:
Merge branch 'master' into master

karthikRavichandran committed:
Merge branch 'master' into master

karthikRavichandran commented:
@hwchase17 @baskaryan . I've made suggested chages and verfied the testcase in local (pylint). Please let me if I need to change anything.

karthikRavichandran committed:
Merge branch 'langchain-ai:master' into master

karthikRavichandran committed:
fix the other errors

karthikRavichandran committed:
fix few pylint errors

karthikRavichandran committed:
fix few pylint errors

karthikRavichandran committed:
fix few pylint errors

karthikRavichandran committed:
fixed some pylint errors

karthikRavichandran committed:
Merge branch 'master' into master

karthikRavichandran committed:
Hope I fixed all the error - let it be final push. God pl help me

karthikRavichandran committed:
Merge branch 'master' into master

karthikRavichandran committed:
lint issue fixed

karthikRavichandran committed:
Merge remote-tracking branch 'origin/master'

karthikRavichandran committed:
Merge branch 'langchain-ai:master' into master

karthikRavichandran committed:
Merge branch 'langchain-ai:master' into master

karthikRavichandran committed:
lint fixed in notebook

karthikRavichandran committed:
Merge remote-tracking branch 'origin/master'

karthikRavichandran committed:
Merge branch 'langchain-ai:master' into master

karthikRavichandran commented:
@baskaryan @hwchase17 I have fixed all CI errors, now all the checks are 💯  done and successful . Can you please verify it and let me know the further changes.

karthikRavichandran committed:
Merge branch 'langchain-ai:master' into master

karthikRavichandran committed:
Merge branch 'master' into master

karthikRavichandran committed:
Merge branch 'master' into master

karthikRavichandran commented:
@hwchase17 @baskaryan , please let me know if you want me to change anything in the code- like class name or function structure. All the test cases are passing as of now

karthikRavichandran committed:
Merge branch 'master' into master

karthikRavichandran commented:
Hi @baskaryan, @efriis, @eyurtsev, @hwchase17, Can any of you please look at my pull request? I get success mail for all the test case but the workflows says it's till waiting for approval

karthikRavichandran committed:
Merge branch 'langchain-ai:master' into master

baskaryan committed:
Merge branch 'master' into master

karthikRavichandran commented:
@baskaryan Do I have to update the PR whenever the master branch changes ? Or Do i have to wait for your review and keep the changes untouched ?.

karthikRavichandran commented:
@baskaryan @hwchase17 Can you please guide me regrading this PR. Do I have to do fix or change anything in this implimentaion?

efriis commented:
closing and feel free to reopen against the langchain-experimental repo (this package moved)! https://github.com/langchain-ai/langchain-experimental

sorry for the delay on this one. This doesn't really fit any of our abstractions well, and thank you for making it a standalone class! I think the remaining issue is that it still mostly implements the retriever interface, and I'm a bit confused why it mostly deals with `Documents` as opposed to a message or strings. But I might be misunderstanding!

efriis commented:
also if you transfer to new repo - please document in the docstring so it shows up in api docs! I don't think adding a notebook to the retrievers integrations section makes sense for this (will show in api docs though)
